8291418: adjust monitor deflation logging and deflate_idle_monitors use

dcubed-ojdk opened the pull request:
Trivial fixes to monitor deflation logging and deflate_idle_monitors use.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8291418](https://bugs.openjdk.org/browse/JDK-8291418): adjust monitor deflation logging and deflate_idle_monitors use


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [f558d870](https://git.openjdk.org/jdk/pull/11293/files/f558d870ea4637fb64f19c6d4deb5346c6032621)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [f558d870](https://git.openjdk.org/jdk/pull/11293/files/f558d870ea4637fb64f19c6d4deb5346c6032621)
 * [Patricio Chilano Mateo](https://openjdk.org/census#pchilanomate) (@pchilano - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11293/head:pull/11293` \
`$ git checkout pull/11293`

Update a local copy of the PR: \
`$ git checkout pull/11293` \
`$ git pull https://git.openjdk.org/jdk pull/11293/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11293`

View PR using the GUI difftool: \
`$ git pr show -t 11293`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11293.diff">https://git.openjdk.org/jdk/pull/11293.diff</a>

</details>

dcubed-ojdk commented:
/label add hotspot-runtime

dholmes-ora reviewed (approved):
This seems reasonable.

Just an observation on the pre-existing logic ... it is interesting that logging has a side-effect of deflating idle monitors. If someone was encountering a problem due to too many inflated monitors and turned on logging to try and show them why, the problem might disappear and the logging would be ineffective in aiding in tracking down the problem.

tstuefe reviewed (approved):
Looks good.

But this felt not trivial to me :) I tried to understand the preexisting code:

`while (ObjectSynchronizer::deflate_idle_monitors(tablep) >= (size_t)MonitorDeflationMax) {`

So, was the intent behind this "as long as deflate_idle_monitors() uses its full allotment of deflations, continue, but if you notice it slacking off, stop?" And the assumption is we always deflate n*MonitorDeflationMax + a smaller rest? And this assumption was wrong?

pchilano commented on src/hotspot/share/runtime/synchronizer.cpp:
```diff
@@ -1548,9 +1548,9 @@ size_t ObjectSynchronizer::deflate_idle_monitors(ObjectMonitorsHashtable* table)
-    if (deflated_count != 0 || log_is_enabled(Debug, monitorinflation)) {
-      ls->print_cr("deflated " SIZE_FORMAT " monitors in %3.7f secs",
-                   deflated_count, timer.seconds());
+    if (deflated_count != 0 || unlinked_count != 0 || deleted_count != 0 || log_is_enabled(Debug, monitorinflation)) {
```
Why check both unlinked_count and deleted_count? Shouldn't they be equal if we got here?

dcubed-ojdk replied:
If memory serves, it's possible for the final safepoint to catch
the MonitorDeflation thread while it's working in here at different
phases of the `deflate_idle_monitors()` work. When the VM Thread
comes in here for the "final audit" it can find different amounts of
work to do in the different phases because the MonitorDeflation
thread might have deflated N monitors, but it wasn't able to unlink
them all. When the VM Thread does its pass, it may deflate fewer
monitors (M), but may find N+M monitors need to be unlinked and
deleted.

Similarly, I think it's possible for the MonitorDeflation thread to
deflate N monitors and unlink N monitors, but then block for the
final safepoint. When the VMThread comes along, it may delete
fewer monitors(M) and unlink fewer monitors(M), but may find
N+M monitors need to be deleted. If the VM Thread's M value is
zero, then it will have `deflated_count == 0`, `unlinked_count == 0`
and `deleted_count != 0`.

pchilano replied:
But isn't the delete_list local to the MonitorDeflation thread? How can the VMThread access those unlinked but not deleted monitors?

dcubed-ojdk replied:
You are right about `delete_list` being local so once the
MonitorDeflation thread has successfully unlinked a deflated
ObjectMonitor and put it on the local `delete_list`, if the
MonitorDeflation thread goes to a final safepoint, then any
ObjectMonitors on the delete_list won't be freed.

It also means that `unlinked_count == deleted_count` in any
thread that reaches L1551. If I change that if-statement, I think
I'd want to add an `assert()` there also.

pchilano reviewed (approved):
Looks good to me.

dcubed-ojdk commented:
@dholmes-ora, @tstuefe and @pchilano - Thanks for the reviews!

Sorry for the delay in getting back to this review. I had an over abundance of
CI/GK work to do before the holiday break and I just finished getting caught
up after the holiday break.

dcubed-ojdk commented:
> This seems reasonable.

Thanks!

> Just an observation on the pre-existing logic ... it is interesting that logging has a side-effect of deflating idle monitors. If someone was encountering a problem due to too many inflated monitors and turned on logging to try and show them why, the problem might disappear and the logging would be ineffective in aiding in tracking down the problem.

I suppose that is a possibility if, for some reason, the MonitorDeflation thread was not
running in a timely fashion (too long of a delay, missed notification, stuck thread).
However, most cases of too many inflated monitors end up being because the monitors
didn't meet the requirements for being deflated (non-zero waiter count being the most
likely).

dcubed-ojdk commented:
> Looks good.

Thanks!


> But this felt not trivial to me :)

Sorry about that. I suspect that I have lived with this code for too long...
I tried to figure out how to update the PR description to put a line thru
the word "Trivial", but I don't see that option.


> I tried to understand the preexisting code:
> 
> `while (ObjectSynchronizer::deflate_idle_monitors(tablep) >= (size_t)MonitorDeflationMax) {`
> 
> So, was the intent behind this "as long as deflate_idle_monitors() uses its full allotment of deflations, continue, but if you notice it slacking off, stop?" And the assumption is we always deflate n*MonitorDeflationMax + a smaller rest? And this assumption was wrong?

My original intent was "loop here until we have nothing left to deflate".
I implemented it wrong and we stopped if there was a partial
deflation, i.e., less than `MonitorDeflationMax` deflated.
However, in some cases, we would see multiple partial deflations
in a row before there is nothing left to deflate so I changed the loop
to stop when we did a complete pass and deflated nothing.

dcubed-ojdk commented:
@pchilano:

> Looks good to me.

Thanks.

dcubed-ojdk commented:
@dholmes-ora, @tstuefe and @pchilano -

I think I've replied to all the review comments. Please let me know if
you are okay with the replies.

tstuefe commented:
All good. Thanks for answering my question.

dcubed-ojdk commented:
Logging output with the latest commit looks like this:
```
[0.040s][info][monitorinflation] Starting the final audit.
[0.040s][info][monitorinflation] begin deflating: in_use_list stats: ceiling=10240, count=2, max=2
[0.040s][info][monitorinflation] deflated_count=1, {unlinked,deleted}_count=1 monitors in 0.0000017 secs
[0.040s][info][monitorinflation] end deflating: in_use_list stats: ceiling=10240, count=1, max=2
```

pchilano reviewed (approved):
Thanks Dan, latest changes look good.

dcubed-ojdk commented:
@pchilano - Thanks for the re-review.

dcubed-ojdk commented:
Local testing of this patch after rebasing to the jdk-20+26 snapshot looks good.

/integrate
